File: frontend/src/lib/system-interfaces.ts
// System Interfaces for Future Implementation
// These interfaces define contracts for DB persistence, caching, model selection, and dev optimizations

// ============================================================================
// DATABASE PERSISTENCE INTERFACES
// ============================================================================

export interface DatabaseConfig {
  type: 'supabase' | 'postgres' | 'mongodb' | 'in-memory'
  connectionString?: string
  schema?: string
  ssl?: boolean
}

export interface UserProfile {
  id: string
  userId: string
  personalitySnapshot?: PersonalitySnapshot
  storyBrain?: StoryBrain
  manualExperience?: ManualExperienceData
  communicationPreferences?: CommunicationPreferences
  onboardingCompleted: boolean
  createdAt: Date
  updatedAt: Date
  lastActiveAt: Date
}

export interface PracticeSession {
  id: string
  userId: string
  question: string
  transcript: string
  audioUrl?: string
  scores: PracticeScores
  feedback: PracticeFeedback
  duration: number
  createdAt: Date
}

export interface StoryBank {
  id: string
  userId: string
  stories: Story[]
  clusters: StoryCluster[]
  embeddingModel: string
  totalStories: number
  createdAt: Date
  updatedAt: Date
}

export interface PracticePlan {
  id: string
  userId: string
  title: string
  description: string
  durationDays: number
  dailyTasks: PracticeTask[]
  targetCompetencies: string[]
  focusAreas: string[]
  progress: PlanProgress
  createdAt: Date
  completedAt?: Date
}

export interface DatabaseAdapter {
  // User Profile Operations
  saveUserProfile(profile: UserProfile): Promise<void>
  getUserProfile(userId: string): Promise<UserProfile | null>
  updateUserProfile(userId: string, updates: Partial<UserProfile>): Promise<void>

  // Practice Sessions
  savePracticeSession(session: PracticeSession): Promise<void>
  getPracticeSessions(userId: string, limit?: number): Promise<PracticeSession[]>
  getPracticeSession(sessionId: string): Promise<PracticeSession | null>

  // Story Bank
  saveStoryBank(storyBank: StoryBank): Promise<void>
  getStoryBank(userId: string): Promise<StoryBank | null>
  updateStoryBank(userId: string, updates: Partial<StoryBank>): Promise<void>

  // Practice Plans
  savePracticePlan(plan: PracticePlan): Promise<void>
  getPracticePlans(userId: string): Promise<PracticePlan[]>
  updatePracticePlan(planId: string, updates: Partial<PracticePlan>): Promise<void>

  // Analytics
  getUserAnalytics(userId: string): Promise<UserAnalytics>
}

// ============================================================================
// CACHING INTERFACES
// ============================================================================

export interface CacheConfig {
  type: 'memory' | 'redis' | 'file' | 'localStorage'
  ttl?: number // Time to live in seconds
  maxSize?: number // Maximum cache size
  namespace?: string
}

export interface CacheEntry<T = any> {
  key: string
  value: T
  expiresAt?: Date
  metadata?: Record<string, any>
}

export interface CacheAdapter {
  get<T>(key: string): Promise<T | null>
  set<T>(key: string, value: T, ttl?: number): Promise<void>
  delete(key: string): Promise<void>
  clear(): Promise<void>
  has(key: string): Promise<boolean>
  keys(pattern?: string): Promise<string[]>
  size(): Promise<number>
}

// Specific cache interfaces for different data types
export interface DemoAnswerCache {
  getCachedAnswer(question: string, context?: AnswerContext): Promise<DemoAnswer | null>
  setCachedAnswer(question: string, answer: DemoAnswer, context?: AnswerContext): Promise<void>
  clearExpired(): Promise<void>
}

export interface StoryGenerationCache {
  getCachedStories(userId: string): Promise<Story[] | null>
  setCachedStories(userId: string, stories: Story[]): Promise<void>
  invalidateUserStories(userId: string): Promise<void>
}

export interface PersonalizationCache {
  getCachedProfile(userId: string): Promise<PersonalizationProfile | null>
  setCachedProfile(userId: string, profile: PersonalizationProfile): Promise<void>
  updatePersonalityTraits(userId: string, traits: PersonalityTraits[]): Promise<void>
}

// ============================================================================
// MODEL SELECTION INTERFACES
// ============================================================================

export type AIModel = 'gemini' | 'claude-sonnet' | 'claude-haiku' | 'gpt-4' | 'gpt-3.5-turbo'

export interface ModelConfig {
  provider: 'google' | 'anthropic' | 'openai'
  model: string
  temperature: number
  maxTokens: number
  costPerToken?: number
  capabilities: ModelCapabilities
}

export interface ModelCapabilities {
  textGeneration: boolean
  audioTranscription: boolean
  embeddings: boolean
  functionCalling: boolean
  vision?: boolean
}

export interface TaskModelMapping {
  demoAnswer: AIModel
  personalityAnalysis: AIModel
  resumeProcessing: AIModel
  storyGeneration: AIModel
  answerPersonalization: AIModel
  practiceScoring: AIModel
  audioTranscription: AIModel
}

export interface ModelSelector {
  selectModelForTask(task: keyof TaskModelMapping): AIModel
  getModelConfig(model: AIModel): ModelConfig
  updateTaskModel(task: keyof TaskModelMapping, model: AIModel): void
  getAvailableModels(): AIModel[]
  validateModelSupport(model: AIModel, task: string): boolean
}

// ============================================================================
// DEV OPTIMIZATION INTERFACES
// ============================================================================

export interface DevConfig {
  enableCaching: boolean
  enableFixtures: boolean
  mockAIResponses: boolean
  skipOnboarding: boolean
  useLocalStorage: boolean
  enableDebugLogging: boolean
  mockAudioRecording: boolean
  fastMode: boolean // Skip delays and animations
}

export interface FixtureData {
  demoAnswers: Record<string, DemoAnswer>
  personalityProfiles: Record<string, PersonalitySnapshot>
  storyBanks: Record<string, Story[]>
  practiceSessions: Record<string, PracticeSession[]>
  resumeSamples: Record<string, string>
}

export interface DevTools {
  loadFixtures(): Promise<void>
  resetAllData(): Promise<void>
  enableFastMode(): void
  disableFastMode(): void
  mockAPIResponse(endpoint: string, response: any): void
  clearMockResponses(): void
  exportUserData(userId: string): Promise<string>
  importUserData(userId: string, data: string): Promise<void>
}

export interface PerformanceMonitor {
  startTiming(label: string): () => number // Returns end function that returns duration
  recordMetric(name: string, value: number, tags?: Record<string, string>): void
  getMetrics(name?: string): PerformanceMetric[]
  clearMetrics(): void
}

export interface PerformanceMetric {
  name: string
  value: number
  timestamp: Date
  tags?: Record<string, string>
}

// ============================================================================
// TYPE DEFINITIONS
// ============================================================================

// Re-export types from existing schemas
export type {
  PersonalityTraits,
  CommunicationStyle,
  PersonalitySnapshot,
  ManualExperienceData,
  Story,
  StoryCluster,
  StoryBrain,
  PracticeScores,
  PracticeFeedback,
  DemoAnswer,
  AnswerContext,
  PersonalizationProfile,
} from './models'

// Additional types for system interfaces
export interface CommunicationPreferences {
  preferredStyle: 'direct' | 'narrative' | 'conversational'
  formalityLevel: 'formal' | 'moderate' | 'casual'
  detailLevel: 'high' | 'balanced' | 'concise'
  pacing: 'slow' | 'moderate' | 'fast'
}

export interface PracticeTask {
  id: string
  title: string
  description: string
  type: 'practice' | 'review' | 'reflection' | 'planning'
  duration: number // minutes
  completed: boolean
  completedAt?: Date
  notes?: string
}

export interface PlanProgress {
  completedTasks: number
  totalTasks: number
  currentDay: number
  streakDays: number
  lastCompletedAt?: Date
}

export interface UserAnalytics {
  totalPracticeSessions: number
  averageScore: number
  improvementTrend: number[]
  mostPracticedCompetencies: string[]
  weakestAreas: string[]
  timeSpentPracticing: number // minutes
  streakData: {
    current: number
    longest: number
    lastBreakDate?: Date
  }
}

// ============================================================================
// IMPLEMENTATION NOTES
// ============================================================================

/**
 * DATABASE MIGRATION PATH:
 * 1. Current: In-memory storage in StorageService
 * 2. Phase 1: Supabase integration with these interfaces
 * 3. Phase 2: Full PostgreSQL with complex queries
 * 4. Phase 3: Redis caching layer
 */

/**
 * CACHE STRATEGY:
 * - Demo answers: Cache by question hash for 24 hours
 * - Story generation: Cache per user until profile changes
 * - Personalization: Cache personality data for session duration
 * - API responses: Cache successful responses for 1 hour
 */

/**
 * MODEL SELECTION LOGIC:
 * - Demo answers: Fast model (Claude Haiku) for speed
 * - Complex analysis: Advanced model (Claude Sonnet) for accuracy
 * - Audio transcription: Specialized model with Whisper API
 * - Embeddings: Optimized model for semantic similarity
 */

/**
 * DEV OPTIMIZATION FEATURES:
 * - Fixture loading for consistent testing
 * - Mock responses for offline development
 * - Fast mode to skip animations/delays
 * - Data export/import for debugging
 * - Performance monitoring for optimization
 */

File: backend/app/core/system_interfaces.py
"""
System Interfaces for Future Implementation
Database persistence, caching, model selection, and dev optimizations
"""
from typing import Dict, List, Any, Optional, Protocol
from datetime import datetime
from dataclasses import dataclass
from enum import Enum

# ============================================================================
# DATABASE INTERFACES
# ============================================================================

class DatabaseType(Enum):
    SUPABASE = "supabase"
    POSTGRESQL = "postgresql"
    MONGODB = "mongodb"
    IN_MEMORY = "in_memory"

@dataclass
class DatabaseConfig:
    type: DatabaseType
    connection_string: Optional[str] = None
    schema: Optional[str] = None
    ssl: bool = True

@dataclass
class UserProfile:
    id: str
    user_id: str
    personality_snapshot: Optional[Dict[str, Any]] = None
    story_brain: Optional[Dict[str, Any]] = None
    manual_experience: Optional[Dict[str, Any]] = None
    onboarding_completed: bool = False
    created_at: datetime = None
    updated_at: datetime = None
    last_active_at: datetime = None

    def __post_init__(self):
        if self.created_at is None:
            self.created_at = datetime.utcnow()
        if self.updated_at is None:
            self.updated_at = datetime.utcnow()
        if self.last_active_at is None:
            self.last_active_at = datetime.utcnow()

@dataclass
class PracticeSession:
    id: str
    user_id: str
    question: str
    transcript: str
    audio_url: Optional[str] = None
    scores: Dict[str, Any] = None
    feedback: Dict[str, Any] = None
    duration: float = 0.0
    created_at: datetime = None

    def __post_init__(self):
        if self.created_at is None:
            self.created_at = datetime.utcnow()
        if self.scores is None:
            self.scores = {}
        if self.feedback is None:
            self.feedback = {}

@dataclass
class StoryBank:
    id: str
    user_id: str
    stories: List[Dict[str, Any]] = None
    clusters: List[Dict[str, Any]] = None
    embedding_model: str = "sentence-transformers"
    total_stories: int = 0
    created_at: datetime = None
    updated_at: datetime = None

    def __post_init__(self):
        if self.created_at is None:
            self.created_at = datetime.utcnow()
        if self.updated_at is None:
            self.updated_at = datetime.utcnow()
        if self.stories is None:
            self.stories = []
        if self.clusters is None:
            self.clusters = []

class DatabaseAdapter(Protocol):
    """Protocol for database adapters"""

    async def save_user_profile(self, profile: UserProfile) -> None:
        """Save user profile"""
        ...

    async def get_user_profile(self, user_id: str) -> Optional[UserProfile]:
        """Get user profile by user_id"""
        ...

    async def update_user_profile(self, user_id: str, updates: Dict[str, Any]) -> None:
        """Update user profile"""
        ...

    async def save_practice_session(self, session: PracticeSession) -> None:
        """Save practice session"""
        ...

    async def get_practice_sessions(self, user_id: str, limit: int = 10) -> List[PracticeSession]:
        """Get user's practice sessions"""
        ...

    async def save_story_bank(self, story_bank: StoryBank) -> None:
        """Save story bank"""
        ...

    async def get_story_bank(self, user_id: str) -> Optional[StoryBank]:
        """Get user's story bank"""
        ...

# ============================================================================
# CACHE INTERFACES
# ============================================================================

class CacheType(Enum):
    MEMORY = "memory"
    REDIS = "redis"
    FILE = "file"
    NONE = "none"

@dataclass
class CacheConfig:
    type: CacheType
    ttl_seconds: int = 3600  # 1 hour default
    max_size: int = 1000
    namespace: str = "behavaced"

@dataclass
class CacheEntry:
    key: str
    value: Any
    expires_at: Optional[datetime] = None
    metadata: Dict[str, Any] = None

    def __post_init__(self):
        if self.metadata is None:
            self.metadata = {}

class CacheAdapter(Protocol):
    """Protocol for cache adapters"""

    async def get(self, key: str) -> Optional[Any]:
        """Get cached value"""
        ...

    async def set(self, key: str, value: Any, ttl: Optional[int] = None) -> None:
        """Set cached value"""
        ...

    async def delete(self, key: str) -> None:
        """Delete cached value"""
        ...

    async def clear(self) -> None:
        """Clear all cached values"""
        ...

    async def has(self, key: str) -> bool:
        """Check if key exists"""
        ...

    async def keys(self, pattern: str = "*") -> List[str]:
        """Get keys matching pattern"""
        ...

    async def size(self) -> int:
        """Get cache size"""
        ...

# ============================================================================
# MODEL SELECTION INTERFACES
# ============================================================================

class AIModel(Enum):
    GEMINI = "gemini"
    CLAUDE_SONNET = "claude-sonnet"
    CLAUDE_HAIKU = "claude-haiku"
    GPT4 = "gpt-4"
    GPT35_TURBO = "gpt-3.5-turbo"

@dataclass
class ModelConfig:
    provider: str  # 'google', 'anthropic', 'openai'
    model_name: str
    temperature: float
    max_tokens: int
    cost_per_token: Optional[float] = None
    supports_audio: bool = False
    supports_embeddings: bool = False
    supports_functions: bool = False

@dataclass
class TaskModelMapping:
    demo_answer: AIModel = AIModel.CLAUDE_HAIKU
    personality_analysis: AIModel = AIModel.CLAUDE_SONNET
    resume_processing: AIModel = AIModel.GEMINI
    story_generation: AIModel = AIModel.CLAUDE_SONNET
    answer_personalization: AIModel = AIModel.CLAUDE_SONNET
    practice_scoring: AIModel = AIModel.GEMINI
    audio_transcription: AIModel = AIModel.GEMINI  # Uses Whisper API

class ModelSelector(Protocol):
    """Protocol for model selection logic"""

    def select_model_for_task(self, task: str) -> AIModel:
        """Select appropriate model for task"""
        ...

    def get_model_config(self, model: AIModel) -> ModelConfig:
        """Get configuration for model"""
        ...

    def update_task_model(self, task: str, model: AIModel) -> None:
        """Update model for specific task"""
        ...

    def get_available_models(self) -> List[AIModel]:
        """Get list of available models"""
        ...

# ============================================================================
# DEV OPTIMIZATION INTERFACES
# ============================================================================

@dataclass
class DevConfig:
    enable_caching: bool = True
    enable_fixtures: bool = False
    mock_ai_responses: bool = False
    skip_onboarding: bool = False
    use_local_cache: bool = True
    enable_debug_logging: bool = False
    mock_audio_recording: bool = False
    fast_mode: bool = False

@dataclass
class FixtureData:
    demo_answers: Dict[str, Dict[str, Any]]
    personality_profiles: Dict[str, Dict[str, Any]]
    story_banks: Dict[str, List[Dict[str, Any]]]
    practice_sessions: Dict[str, List[Dict[str, Any]]]
    resume_samples: Dict[str, str]

class DevTools(Protocol):
    """Protocol for development tools"""

    def load_fixtures(self) -> None:
        """Load fixture data for testing"""
        ...

    def reset_all_data(self) -> None:
        """Reset all application data"""
        ...

    def enable_fast_mode(self) -> None:
        """Enable fast mode (skip delays)"""
        ...

    def disable_fast_mode(self) -> None:
        """Disable fast mode"""
        ...

    def mock_api_response(self, endpoint: str, response: Dict[str, Any]) -> None:
        """Mock API response for testing"""
        ...

    def clear_mock_responses(self) -> None:
        """Clear all mocked responses"""
        ...

@dataclass
class PerformanceMetric:
    name: str
    value: float
    timestamp: datetime
    tags: Dict[str, str] = None

    def __post_init__(self):
        if self.tags is None:
            self.tags = {}

class PerformanceMonitor(Protocol):
    """Protocol for performance monitoring"""

    def start_timing(self, label: str) -> callable:
        """Start timing operation, returns end function"""
        ...

    def record_metric(self, name: str, value: float, tags: Optional[Dict[str, str]] = None) -> None:
        """Record performance metric"""
        ...

    def get_metrics(self, name: Optional[str] = None) -> List[PerformanceMetric]:
        """Get recorded metrics"""
        ...

    def clear_metrics(self) -> None:
        """Clear all metrics"""
        ...

# ============================================================================
# CONFIGURATION
# ============================================================================

# Default model configurations
MODEL_CONFIGS = {
    AIModel.GEMINI: ModelConfig(
        provider="google",
        model_name="gemini-2.5-flash",
        temperature=0.7,
        max_tokens=8192,
        supports_embeddings=True,
        supports_functions=False
    ),
    AIModel.CLAUDE_SONNET: ModelConfig(
        provider="anthropic",
        model_name="claude-3-5-sonnet-20241022",
        temperature=0.7,
        max_tokens=4096,
        cost_per_token=0.000015,
        supports_functions=True
    ),
    AIModel.CLAUDE_HAIKU: ModelConfig(
        provider="anthropic",
        model_name="claude-3-5-haiku-20241022",
        temperature=0.3,
        max_tokens=4096,
        cost_per_token=0.0000025,
        supports_functions=True
    ),
    AIModel.GPT4: ModelConfig(
        provider="openai",
        model_name="gpt-4",
        temperature=0.7,
        max_tokens=4096,
        cost_per_token=0.00003,
        supports_functions=True
    ),
    AIModel.GPT35_TURBO: ModelConfig(
        provider="openai",
        model_name="gpt-3.5-turbo",
        temperature=0.7,
        max_tokens=4096,
        cost_per_token=0.000002,
        supports_functions=True
    )
}

# Default task-to-model mapping
DEFAULT_TASK_MAPPING = TaskModelMapping()

# ============================================================================
# MIGRATION NOTES
# ============================================================================

"""
DATABASE MIGRATION PATH:
1. Current: In-memory storage in StorageService
2. Phase 1: Supabase integration with these interfaces
3. Phase 2: Full PostgreSQL with complex queries and indexing
4. Phase 3: Redis caching layer for performance

CACHE STRATEGY:
- Demo answers: Cache by question hash for 24 hours
- Story generation: Cache per user until profile changes
- Personalization: Cache personality data for session duration
- API responses: Cache successful responses for 1 hour

MODEL SELECTION LOGIC:
- Demo answers: Fast model (Claude Haiku) for speed
- Complex analysis: Advanced model (Claude Sonnet) for accuracy
- Audio transcription: Specialized model with Whisper API
- Embeddings: Optimized model for semantic similarity

DEV OPTIMIZATION FEATURES:
- Fixture loading for consistent testing
- Mock responses for offline development
- Fast mode to skip animations/delays
- Data export/import for debugging
- Performance monitoring for optimization
"""
